lib/appmap/event.rb

```diff
@@ -60,16 +60,37 @@ def display_string(value)
           final ? value_string : encode_display_string(value_string)
         end
 
-        def object_properties(hash_like)
-          hash = hash_like.to_h
-          hash.keys.map do |key|
-            {
+        def add_schema(param, value, always: false)
+          return unless always || AppMap.parameter_schema?
+
+          if value.blank? || value.is_a?(String)
+            # pass
+          elsif value.is_a?(Enumerable)
+            if value.is_a?(Hash)
+              param[:properties] = object_properties(value)
+            elsif value.respond_to?(:first) && value.first
+              param[:properties] = object_properties(value.first)
+            end
+          elsif value.respond_to?(:as_json)
+            add_schema param, JSON.parse(value.to_json), always: always
+          end
+        end
+
+        def object_properties(hash)
+          hash = hash.attributes if hash.respond_to?(:attributes)
+          hash = hash.to_h if hash.is_a?(Struct)
```

Comment:
Why not do `.to_h` unconditionally at this point? That way it would work with more things than just hashes and structs.

Reply:
I will give it a try.

Reply:
There are various types that respond to `to_h` but blow up if you try and use it. I'm not sure how to address it. Here are some examples:

```
Array#to_h failed: wrong element type Set at 0 (expected array)
Set#to_h failed: wrong element type String (expected array)
Psych::Nodes::Scalar#to_h failed: wrong element type Psych::Nodes::Scalar (expected array)
```